tedana/selection/selection_utils.py

```diff
@@ -313,19 +320,137 @@ def clean_dataframe(component_table):
 #################################################
 
 
-def confirm_metrics_exist(component_table, necessary_metrics, function_name=None):
+def expand_dict(node: Dict, field: str, metrics: List[str]) -> Tuple[bool, List[Dict]]:
+    """Expand a dictionary with regular expressions.
+
+    Parameters
+    ----------
+    node : dict
+        A dictionary containing nested dictionaries called "parameters" and,
+        optionally, "kwargs".
+        Any of the values in the "parameters" or "kwargs" dictionaries may be
+        a regular expression, denoted by starting with "^".
+    field : :obj:`str`
+        The key in the ``node`` dictionary that should be expanded.
+        Current keys that could include regular expressions are "parameters" and "kwargs"
+    metrics : list of str
+        List of metric names to compare regular expressions against.
+
+    Returns
+    -------
+    regex_found : :obj:`bool`
+        True if any regular expressions were found in the input dictionary
+    out_nodes : :obj:`list[dict]`
+        A list of dictionaries with regular expressions replaced by all
+        matching values in the 'metrics' list.
+    """
+    regex_found = False
+    out_nodes = []
+    for k, v in node.get(field, {}).items():
+        if isinstance(v, str) and v.startswith("^"):
+            regex_found = True
+            replacements = [metric for metric in metrics if re.match(v, metric)]
+            if not replacements:
+                raise ValueError(f"No metrics matching regex '{v}' found.")
+
+            for replacement in replacements:
+                LGR.warning(f"Replacing {v} with {replacement}")
+                mod_node = copy.deepcopy(node)
+                mod_node[field][k] = replacement
+                out_nodes.append(mod_node)
+
+    return regex_found, out_nodes
+
+
+def expand_node(node: Dict, metrics: List[str]):
+    """Expand node definitions with regular expressions.
+
+    Recursively expand a node so that any regular expressions are replaced with
+    any matching values in the 'metrics' list.
+    Regular expressions may be present as the value of any key in the
+    subdictionaries "parameters" or "kwargs".
+
+    Parameters
+    ----------
+    node : :obj:`dict`
+        A dictionary containing nested dictionaries called "parameters" and,
+        optionally, "kwargs".
+        Any of the values in the "parameters" or "kwargs" dictionaries may be
+        a regular expression, denoted by starting with "^".
+    metrics : :obj:`list[str]`
+        List of metric names to compare regular expressions against.
+
+    Returns
+    -------
+    out_nodes or expanded_out_nodes: :obj:`list[dict]`
+        A list of dictionaries with regular expressions replaced by all
+        matching values in the 'metrics' list.
+    """
+    regex_found, out_nodes = expand_dict(node, "parameters", metrics)
+    if not regex_found:
```

Comment:
Should this conditional be removed? `expand_dict` is first run for "parameters". Shouldn't it also be run for "kwargs" whether or not regular expressed were expanded for "parameters"

Reply:
Maybe I'm misremembering what it does, but it seems like `expand_dict` could be removed entirely.

Reply:
I'm looking into this more and I don't think we're using `expand_nodes`, `expand_node`, or `expand_dict` anymore. They are designed if someone uses a regular expression as a parameter for a `selection_node` function. I think you used that in a new selection_node function, but that part of the code was removed when we removed correlations. I don't think this is possible for any of our current selection node functions since I don't think any of them can takes more than one metric in a single parameter field. I ran integration tests without these functions, and it was fine.

I'm expanding the regular expressions I am using in [`validate_extern_regress`](https://github.com/ME-ICA/tedana/blob/c26b86d97438a78f7bab7e3d99d9cfb594f8251f/tedana/metrics/external.py#L62).

Reply:
It sounds like you can completely remove those functions then. I'm totally fine with that.